packages/platform/src/setup/setupAccounts.ts

```diff
@@ -71,6 +71,115 @@ export const setupAccounts = (unchainedAPI: UnchainedCore) => {
     },
   };
 
+  accountsServer.services.oauth2 = {
+    async authenticate({
+      authorizationCode,
+      provider,
+      redirectURL,
+    }: {
+      authorizationCode: any;
+      provider: string;
+      redirectURL: string;
+    }): Promise<any> {
+      if (!authorizationCode || !provider) {
+        return undefined;
+      }
+      const { services, modules } = unchainedAPI;
+
+      const oauth2Service = await services.accounts.oauth2({ provider, redirectURL }, unchainedAPI);
+
+      const userAuthorizationToken = await oauth2Service.getAuthorizationCode(authorizationCode);
+      console.log(userAuthorizationToken);
+
+      if (!userAuthorizationToken) throw new Error('Unable to authorize user');
+
+      const userOAuthInfo = await oauth2Service.getAccountData(userAuthorizationToken);
+      console.log(userOAuthInfo);
+
+      if (!userOAuthInfo || !userOAuthInfo?.email) {
+        throw new Error('OAuth authentication failed');
+      }
+
+      try {
+        const user = await unchainedAPI.modules.users.findUser({
```

Comment:
I thought a lot about this. We shouldn't go for an email match, that could be different users and also users could get disconnected if email is changed on provider side. 

When signing in with an oAuth provider, we should get the unique userId from the provider and check if there is already a user that has that id in "user.services.oauth.PROVIDER". If there is a match, that's the user we are looking for and there we should store the current authorizationToken (and eventually refreshToken).

If there is no match then we should try to register the user with the mapped registration data, email and username. If that doesn't work because the email or username is already used by some other user, cancel the registration (in that case we show something like that to the user: "The E-Mail address or username returned from OAUTH_PROVIDER_XYZ already matches a user in our system, please log in with that user first, you can link it with OAUTH_PROVIDER_XYZ throught the account page after)."

Then let's look at the the situation when somebody created a user with email mike@mike.com and password or mike with WebAuthn before. In those cases the user should be able to "link" his social account on an account page, so we need some mutations for that ("linkWithOAuth" "unlinkFromOAuth") or something like that.


Look at other pages like "dev.to" for example:
<img width="1205" alt="Bildschirm­foto 2023-03-07 um 08 10 38" src="https://user-images.githubusercontent.com/2109393/223349871-7dcfc6f4-c9b7-45b5-9ff3-914249554ea4.png">